File: data/python/tests/test_tenacity_wait_random_exponential.py
import unittest
from function_tenacity_wait_random_exponential import RetryFunction
from tenacity import RetryError
import time
from unittest.mock import patch

class TestMightFailFunction(unittest.TestCase):

    def setUp(self):
        self.func = RetryFunction()

    def test_success_after_retry(self):
        result = self.func.might_fail_function(5)
        self.assertEqual(result, 5)

    def test_zero_input(self):
        with self.assertRaises(RetryError) as context:
            self.func.might_fail_function(0)
        self.assertIsInstance(context.exception.last_attempt.exception(), ValueError)

    def test_retry_with_random_exponential_backoff(self):
        start_time = time.time()
        with self.assertRaises(RetryError):
            self.func.might_fail_function(2, extra_sleep=2)
        end_time = time.time()
        total_time = end_time - start_time
        print(f"Test execution time: {total_time:.2f} seconds")
        self.assertGreater(total_time, 6) 
        self.assertLess(total_time, 30)

    def test_success_on_first_try(self):
        result = self.func.might_fail_function(6)
        self.assertEqual(result, 6)

    def test_failure_after_max_attempts(self):
        with self.assertRaises(RetryError) as context:
            self.func.might_fail_function(2)
        self.assertIsInstance(context.exception.last_attempt.exception(), ValueError)

if __name__ == '__main__': 
    unittest.main()

File: data/cpp/tests/test_tenacity_wait_random_exponential.cpp
#include <gtest/gtest.h>
#include "../src/function_tenacity_wait_random_exponential.cpp"


class TestMightFailFunction : public ::testing::Test {
protected:
    RetryFunction func;
};

TEST_F(TestMightFailFunction, TestSuccessAfterRetry) {
    int result = func.might_fail_function(5);
    EXPECT_EQ(result, 5);
}

TEST_F(TestMightFailFunction, TestZeroInput) {
    EXPECT_THROW({
        func.might_fail_function(0);
    }, std::runtime_error);
}

TEST_F(TestMightFailFunction, TestRetryWithRandomExponentialBackoff) {
    auto start_time = std::chrono::high_resolution_clock::now();
    EXPECT_THROW({
        func.might_fail_function(2, 2);
    }, std::runtime_error);
    auto end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> total_time = end_time - start_time;
    EXPECT_GT(total_time.count(), 6);
    EXPECT_LT(total_time.count(), 30);
}

TEST_F(TestMightFailFunction, TestSuccessOnFirstTry) {
    int result = func.might_fail_function(6);
    EXPECT_EQ(result, 6);
}

TEST_F(TestMightFailFunction, TestFailureAfterMaxAttempts) {
    EXPECT_THROW({
        func.might_fail_function(2);
    }, std::runtime_error);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}